### ISSUE TITLE
sit tunnel with Local=dhcp4 fails to update local address when underlying interface gets a new DHCPv4 lease

### ISSUE THREAD

**Issue by kpfleming** [systemd/systemd#24854]
### systemd version the issue has been seen with

systemd 251.4-3

### Used distribution

Debian testing (Bookworm)

### Linux kernel version used

5.19.0-1-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

The `sit` tunnel should have updated its 'Local' address to the new address that the underlying interface obtained.

### Unexpected behaviour you saw

The `sit` tunnel continued listing an old DHCPv4-obtained address, even after repeated up/down cycles of the tunnel.

### Steps to reproduce the problem

This will be challenging to reproduce as I've got on a production system, but here goes:

fios.network
```
[Match]
Name=fios

[Link]
ActivationPolicy=manual

[Network]
LinkLocalAddressing=no
IPForward=yes
DHCP=ipv4
Tunnel=he
```

he.netdev
```
[NetDev]
Name=he
Kind=sit
MTUBytes=1480

[Tunnel]
Local=dhcp4
Remote=209.51.161.14
TTL=255
```

he.network
```
[Match]
Name=he

[Network]
Address=2001:470:1f06:fab::2/64
Gateway=2001:470:1f06:fab::1
IPForward=yes
```

1. Brought the 'fios' interface up using 'ip link set up fios'.
2. DHCPv4 lease was obtained from ISP (log below).
3. 'he' SIT tunnel showed that address as its 'Local' address in 'ip addr ls'.
4. IPv6 traffic flowed as expected over the tunnel.
5. Brought the 'fios' interface down using 'ip link set down fios'.
(days pass)
1. Brought the 'fios' interface up using 'ip link set up fios'.
2. DHCPv4 lease was obtained from ISP (log below).
3. 'he' SIT tunnel showed the *old* address as its 'Local' address in 'ip addr ls'.
4. IPv6 traffic would not flow as the local address was incorrect.
5. Tried various networkctl commands, including down/up of the 'he' link. Address did not change. Even 'systemctl restart systemd-networkd' was insufficient.
6. Resorted to rebooting the system, when the 'fios' interface got a (third) address, the 'he' interface configured itself to use that as 'Local' and IPv6 traffic began flowing.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Sep 25 19:43:30 edge21-a systemd-networkd[867]: fios: Link UP
Sep 25 19:43:30 edge21-a systemd-networkd[867]: fios: Gained carrier
Sep 25 19:43:31 edge21-a systemd-networkd[867]: fios: DHCPv4 address 68.132.194.97/24, gateway 68.132.194.1 acquired from 68.132.194.1
Sep 25 19:43:31 edge21-a systemd-networkd[867]: he: netdev ready
Sep 25 19:43:31 edge21-a systemd-networkd[867]: he: Configuring with /etc/systemd/network/he.network.
Sep 25 19:43:31 edge21-a systemd-networkd[867]: he: Link UP
Sep 25 19:43:31 edge21-a systemd-networkd[867]: he: Gained carrier
Sep 26 17:26:33 edge21-a systemd-networkd[867]: ocftest: netdev ready
Sep 26 17:26:33 edge21-a systemd-networkd[867]: ocftest: Configuring with /etc/systemd/network/ocftest.network.
Sep 26 20:42:09 edge21-a systemd-networkd[867]: ocftest: Link UP
Sep 26 20:42:09 edge21-a systemd-networkd[867]: ocftest: Gained carrier
Sep 26 20:42:09 edge21-a systemd-networkd[867]: ocftest: Gained IPv6LL
Sep 26 20:47:48 edge21-a systemd-networkd[867]: ocftest: Link DOWN
Sep 26 20:47:48 edge21-a systemd-networkd[867]: ocftest: Lost carrier
Sep 26 20:47:48 edge21-a systemd-networkd[867]: ocftest: DHCPv6 lease lost
Sep 26 20:48:48 edge21-a systemd-networkd[867]: ocftest: Link UP
Sep 26 20:48:48 edge21-a systemd-networkd[867]: ocftest: Gained carrier
Sep 26 20:48:48 edge21-a systemd-networkd[867]: ocftest: Gained IPv6LL
Sep 26 20:49:13 edge21-a systemd-networkd[867]: ocftest: Link DOWN
Sep 26 20:49:13 edge21-a systemd-networkd[867]: ocftest: Lost carrier
Sep 26 20:49:13 edge21-a systemd-networkd[867]: ocftest: DHCPv6 lease lost
Sep 27 20:24:36 edge21-a systemd-networkd[867]: tailscale0: Link DOWN
Sep 27 20:24:36 edge21-a systemd-networkd[867]: tailscale0: Lost carrier
Sep 27 20:25:32 edge21-a systemd-networkd[867]: tailscale0: Link UP
Sep 27 20:25:32 edge21-a systemd-networkd[867]: tailscale0: Gained carrier
Sep 27 20:25:32 edge21-a systemd-networkd[867]: tailscale0: Gained IPv6LL
Sep 27 20:30:51 edge21-a systemd-networkd[867]: tailscale0: Link DOWN
Sep 27 20:30:51 edge21-a systemd-networkd[867]: tailscale0: Lost carrier
Sep 27 20:30:53 edge21-a systemd-networkd[867]: fios: Link DOWN
Sep 27 20:30:53 edge21-a systemd-networkd[867]: fios: Lost carrier
Sep 27 20:30:53 edge21-a systemd-networkd[867]: fios: DHCP lease lost
Sep 28 15:05:38 edge21-a systemd-networkd[867]: ocftest: Unmanaging interface.
Sep 28 15:05:38 edge21-a systemd-networkd[867]: ocftest: DHCPv6 lease lost
Sep 28 17:46:27 edge21-a systemd-networkd[867]: fios: Link UP
Sep 28 17:46:27 edge21-a systemd-networkd[867]: fios: Gained carrier
Sep 28 17:46:31 edge21-a systemd-networkd[867]: fios: DHCPv4 address 108.21.80.74/24, gateway 108.21.80.1 acquired from 108.21.80.1
Sep 28 17:46:31 edge21-a systemd-networkd[867]: tailscale0: Link UP
Sep 28 17:46:31 edge21-a systemd-networkd[867]: tailscale0: Gained carrier
Sep 28 17:46:31 edge21-a systemd-networkd[867]: tailscale0: Gained IPv6LL
Sep 28 17:48:33 edge21-a systemd-networkd[867]: he: Configuring with /etc/systemd/network/he.network.
Sep 28 17:49:42 edge21-a systemd-networkd[867]: he: Link DOWN
Sep 28 17:49:42 edge21-a systemd-networkd[867]: he: Lost carrier
Sep 28 17:49:47 edge21-a systemd-networkd[867]: he: Link UP
Sep 28 17:49:47 edge21-a systemd-networkd[867]: he: Gained carrier
Sep 28 17:51:13 edge21-a systemd[1]: Stopping Network Configuration...
Sep 28 17:51:14 edge21-a systemd-networkd[867]: mgmt: DHCPv6 lease lost
Sep 28 17:51:14 edge21-a systemd-networkd[867]: dns: DHCPv6 lease lost
Sep 28 17:51:14 edge21-a systemd-networkd[867]: ntp: DHCPv6 lease lost
```

**Comment by kpfleming** [comment]
This is now (mostly) easily reproducable. With the configuration above, bring up the links.

Then, `touch /etc/systemd/network/fios.network`, followed by `networkctl reload`.

systemd-networkd will release the DHCP lease and obtain a new one; if the new address differs from the old one, the `sit` tunnel does not update to use it.

**Comment by yuwata** [comment]
Fix is waiting in #25223. If possible, please test the PR. Thank you.

**Comment by kpfleming** [comment]
I've applied the patches to the Debian 252-2 sources and rebuilt, and then installed on one of my machines with debug logging enabled for systemd-networkd.service. Unfortunately when I was able to trigger a DHCP address change on `fios` I didn't see any log messages from `hetunnel` in the systemd-networkd journal, and the local address for the tunnel did not change.

I can try adding some debug logging in the sending side to see if it is even attempting to notify the tunnel interface to update its local address...
